### PR TITLE
Introduce SUBROUTINE GWspatialParameters

### PR DESCRIPTION
**Pull request by rkutteh**
# CABLE ground water hydrology work of Mengyuan Mu

This pull request is part of integrating the ground water hydrology work of Mengyuan Mu into CABLE, which will take place in a sequence of pull requests. The actual integration project itself was completed back in September 2023. The current work involves just a gradual merge into the main branch. 

## Description

1. A new SUBROUTINE GWspatialParameters is introduced in **src/offline/cable_parameters.F90**
2. This subroutine will eventually be called only once, from **src/offline/cable_input.F90**. I will add this call at the appropriate time, but for now the subroutine is inactive
3. Dependencies are introduced in **src/offline/cable_parameters.F90** and **src/offline/cable_define_types.F90** in order for the code to compile successfully (confirmed)
4. This pull request involves only code additions, no deletions or modifications of existing code were carried out (some minor typos I came across in existing comments were fixed)
5. *PLEASE* do not delete any of my in-code comments (tagged rk4417) for now. Once the integration of the ground water hydrology work is complete and its functionality is confirmed to reproduce the results obtained already outside of the repo, I will clean-up all of my comments in a single pull request. Deleting any of my comments now will just make it much more difficult for me to track down and fix any issues that might arise during this process.

----
📚 Documentation preview 📚: https://cable--224.org.readthedocs.build/en/224/

### CODE REVIEW

**Comment by ccarouge** [comment]
@rkutteh I know you were hoping for a quick pull request. But I don't think I can give you that. From what I've seen during the review the GWspatialParameters add to the spaghetti mess of cable_parameters.F90 instead of helping untangling things. So I'm not particularly inclined to accept as is. I'll have a deeper look into it and see if we can adapt things quickly to get the code to a better standard.

**Comment by rkutteh** [comment]
@ccarouge As you might guess from my earlier comments, I prefer that all clean-up of code occurs **after** all the GW work is in and validated against the copy I have outside the repo. If you change things around while I am moving code in and the final code does not agree with my results, it will be nearly impossible for me to track down the source(s) of the errors. As it stands, we know that the code that I am introducing works with GW on, so my own strategy would be to mirror it in the repo, check that all is good as expected, and only THEN clean-up and check, clean-up and check,...Keep in mind that GWspatialParameters is presently inactive.

File: src/offline/cable_parameters.F90 [hunk 1]
```diff
@@ -2740,11 +2756,660 @@ SUBROUTINE init_veg_from_vegin(ifmp,fmp, veg, soil_zse )
...
+         integer, intent(in)               :: ncfile_id
+         real(r_2), dimension(mp) :: data_vec
+
+         real, dimension(nlon,nlat) :: GW2d_data
+         real, dimension(nlon,nlat) :: default_data
+         integer :: i,j,k,n
+         integer :: varid
+         integer :: varinq_status,varget_status
+
+         print *, "MMY in get_gw_2d_var_constdef" ! MMY add to test whether get_gw_2d_var_constdef is necessary
```
Comment (ccarouge) [inline src/offline/cable_parameters.F90:3084]:
```suggestion
```

File: src/offline/cable_parameters.F90 [hunk 2]
```diff
@@ -2740,11 +2756,660 @@ SUBROUTINE init_veg_from_vegin(ifmp,fmp, veg, soil_zse )
...
+         default_data(:,:) = default_const
+
+         if (.not.cable_user%gw_model) then
+
+            GW2d_data(:,:) = default_data(:,:)
+
+         else
```
Comment (ccarouge) [inline src/offline/cable_parameters.F90:3091]: I would like to remove this case. This should be done outside these routines. Keep these generic.

File: src/offline/cable_parameters.F90 [hunk 3]
```diff
@@ -2740,11 +2756,660 @@ SUBROUTINE init_veg_from_vegin(ifmp,fmp, veg, soil_zse )
...
+         real, dimension(nlon,nlat) :: GW2d_data
+         integer :: i,j,k,n
+         integer :: varid
+         integer :: varinq_status,varget_status
+
+         if (.not.cable_user%gw_model) then
+
+            GW2d_data(:,:) = default_data(:,:)
+
+         else
```
Comment (ccarouge) [inline src/offline/cable_parameters.F90:3139]: I would like to remove this case. This should be done outside these routines. Keep these generic.

File: src/offline/cable_parameters.F90 [hunk 4]
```diff
@@ -2740,11 +2756,660 @@ SUBROUTINE init_veg_from_vegin(ifmp,fmp, veg, soil_zse )
...
+         real, dimension(nlon,nlat) :: default_data
+
+          default_data(:,:) = default_const
+
+         if (.not. cable_user%gw_model) then
+            do k=1,ms
+               GW3d_data(:,:,k) = default_data(:,:)
+            end do
+
+         else
```
Comment (ccarouge) [inline src/offline/cable_parameters.F90:3191]: I would like to remove this case. This should be done outside these routines. Keep these generic.

File: src/offline/cable_parameters.F90 [hunk 5]
```diff
@@ -2740,11 +2756,660 @@ SUBROUTINE init_veg_from_vegin(ifmp,fmp, veg, soil_zse )
...
+         integer :: i,j,k,n
+         integer :: varid
+         integer :: varinq_status,varget_status
+
+         if (.not. cable_user%gw_model) then
+            do k=1,ms
+               GW3d_data(:,:,k) = default_data(:,:)
+            end do
+
+         else
```
Comment (ccarouge) [inline src/offline/cable_parameters.F90:3245]: I would like to remove this case. This should be done outside these routines. Keep these generic.

File: src/offline/cable_parameters.F90 [hunk 6]
```diff
@@ -2740,11 +2756,660 @@ SUBROUTINE init_veg_from_vegin(ifmp,fmp, veg, soil_zse )
...
+
+         default_data(:,:) = default_const
+
+         if (.not.cable_user%gw_model) then
+            do k=1,ms
+               do n=1,npatch
+                  GW4d_data(:,:,k,n) = default_data(:,:)
+               end do
+            end do
+         else
```
Comment (ccarouge) [inline src/offline/cable_parameters.F90:3304]: I would like to remove this case. This should be done outside these routines. Keep these generic.

File: src/offline/cable_parameters.F90 [hunk 7]
```diff
@@ -2740,11 +2756,660 @@ SUBROUTINE init_veg_from_vegin(ifmp,fmp, veg, soil_zse )
...
+         real, dimension(nlon,nlat,ms)        :: GW3d_data
+
+
+         if (.not.cable_user%gw_model) then
+            do k=1,ms
+               do n=1,npatch
+                  GW4d_data(:,:,k,n) = default_data(:,:)
+               end do
+            end do
+         else
```
Comment (ccarouge) [inline src/offline/cable_parameters.F90:3370]: I would like to remove this case. This should be done outside these routines. Keep these generic.

**Review by ccarouge — APPROVED** [review]
I have written down what I would change (at least what I could see here). 
But I have seen that `cable_parameters.F90` already has code to read in the GW parameters and assign the values to the CABLE variables (in spatialSoil() and write_default_params()), so we need to ensure there is only one way to do that in the final code.

This clean up will be done separately once the GW code is committed in to ensure testing is possible.

**Comment by rkutteh** [comment]
@ccarouge thanks for all the comments. Some I suspect will be resolved "automatically", once I am finished pushing all the remaining changes to src/offline/cable_parameters.F90 . The remaining I agree we can address up after testing.

**Pull request merged**